pkg/helpers/klusterlet/klusterlet.go

```diff
@@ -0,0 +1,108 @@
+	"fmt"
+	"os"
+
+	"github.com/ghodss/yaml"
+	"k8s.io/klog/v2"
+	operatorv1 "open-cluster-management.io/api/operator/v1"
+	"open-cluster-management.io/ocm/pkg/operator/helpers/chart"
+)
+
+// MergeKlusterletFileJoin merges the klusterlet file into the klusterlet chart config for a klusterlet join.
+// All fields are allowed to be overridden during a klusterlet join.
+func MergeKlusterletFileJoin(klusterletFile string, klusterletChartConfig *chart.KlusterletChartConfig) error {
+	spec, err := loadKlusterletFile(klusterletFile)
+	if err != nil {
+		return fmt.Errorf("failed to load klusterlet file %s: %v", klusterletFile, err)
+	}
+
+	if err := mergeKlusterletFileUpgrade(spec, klusterletChartConfig); err != nil {
+		return fmt.Errorf("failed to merge klusterlet file %s: %v", klusterletFile, err)
+	}
+	if err := mergeKlusterletFileJoin(spec, klusterletChartConfig); err != nil {
+		return fmt.Errorf("failed to merge klusterlet file %s: %v", klusterletFile, err)
+	}
+
+	klog.V(2).InfoS("Successfully merged klusterlet file into klusterlet chart config for join", "file", klusterletFile)
+	return nil
+}
+
+// MergeKlusterletFileUpgrade merges the klusterlet file into the klusterlet chart config for a klusterlet upgrade.
+// Overriding the cluster name, klusterlet namespace, and priority class name is not allowed during upgrades.
+func MergeKlusterletFileUpgrade(klusterletFile string, klusterletChartConfig *chart.KlusterletChartConfig) error {
+	spec, err := loadKlusterletFile(klusterletFile)
+	if err != nil {
+		return fmt.Errorf("failed to load klusterlet file %s: %v", klusterletFile, err)
+	}
+
+	if err := mergeKlusterletFileUpgrade(spec, klusterletChartConfig); err != nil {
+		return fmt.Errorf("failed to merge klusterlet file %s: %v", klusterletFile, err)
+	}
+
+	klog.V(2).InfoS("Successfully merged klusterlet file into klusterlet chart config for upgrade", "file", klusterletFile)
+	return nil
+}
+
+func loadKlusterletFile(klusterletFile string) (*operatorv1.KlusterletSpec, error) {
+	yamlBytes, err := os.ReadFile(klusterletFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read klusterlet file %s: %v", klusterletFile, err)
+	}
+	var klusterlet operatorv1.Klusterlet
+	if err := yaml.Unmarshal(yamlBytes, &klusterlet); err != nil {
+		return nil, fmt.Errorf("failed to parse klusterlet YAML: %v", err)
+	}
+	return &klusterlet.Spec, nil
+}
+
+func mergeKlusterletFileJoin(spec *operatorv1.KlusterletSpec, klusterletChartConfig *chart.KlusterletChartConfig) error {
+	if spec.Namespace != "" {
+		klusterletChartConfig.Klusterlet.Namespace = spec.Namespace
+	}
+	if spec.ClusterName != "" {
+		klusterletChartConfig.Klusterlet.ClusterName = spec.ClusterName
+	}
+	if spec.PriorityClassName != "" {
+		klusterletChartConfig.PriorityClassName = spec.PriorityClassName
+	}
+	return nil
+}
+
+func mergeKlusterletFileUpgrade(spec *operatorv1.KlusterletSpec, klusterletChartConfig *chart.KlusterletChartConfig) error {
+	if spec.DeployOption.Mode != "" {
```

Comment:
/cc @xuezhaojun
If I remember correctly, you have implemented some merge func like this using Go's reflect`, which does not require comparing every field. Do you think we can borrow that here?

Reply:
I guess we mean this https://github.com/xuezhaojun/cluster-lifecycle-api/blob/278b42dcc7df7250efb3748a9a5b03b028e8a08e/helpers/klusterletconfig/merge.go#L12-L28 file.

I doesn't think they are the same case, in this one, 2 structs are different types, I prefer keep it in this simple, one case by one case way.